### ISSUE TITLE
The *.astype method output array loses the maskna property. (migrated from Trac #1973)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3521]
Original ticket http://projects.scipy.org/numpy/ticket/1973
Reported 2011-11-04 by atmention:bsouthey, assigned to atmention:pierregm.

I get an error when displaying a masked array that contains the new missing value, np.NA under 64-bit Linux. But the operations seem to work as expected.

```
>>> import numpy as np
>>> np.__version__
'2.0.0.dev-7297785'
>>> b=np.ma.array([3,5,np.NA,6], mask=[0,0,1,0])
>>> b
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python2.7/site-packages/numpy/ma/core.py", line 3551, in __repr__
    data=str(self), mask=str(self._mask),
  File "/usr/lib64/python2.7/site-packages/numpy/ma/core.py", line 3534, in __str__
    res = self._data.astype("O")
ValueError: Cannot assign NA to an array which does not support NAs
>>> b._data
array([3, 5, NA, 6])
>>> b.sum()
14
>>> b.mask=[0,0,0,0]
>>> b
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python2.7/site-packages/numpy/ma/core.py", line 3551, in __repr__
    data=str(self), mask=str(self._mask),
  File "/usr/lib64/python2.7/site-packages/numpy/ma/core.py", line 3534, in __str__
    res = self._data.astype("O")
ValueError: Cannot assign NA to an array which does not support NAs
>>> b.sum()
NA(dtype='int64')
```

**Comment by thouis** [comment]
Comment in Trac by atmention:bsouthey, 2011-11-04

Sorry I was too quick on the submit button as I tracked this down to the 'astype' function that does not support the new np.NA value. I have no problems if this ticket is closed and another opened for astype. 

```
>>> np.array([1,2,3,4]).astype(float)
array([ 1.,  2.,  3.,  4.])
>>> np.array([1,2,3,np.NA]).astype(float)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Cannot assign NA to an array which does not support NAs
>>> np.array([1,2,3,np.NA]).astype(np.int64)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Cannot assign NA to an array which does not support NAs
```

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2012-01-07

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-01-15

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-01-15

This is fixed in pull request 183.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-01-21

I pushed the fix in 55472ca..6666eec. I'll open another ticket for the real/imag parts problem.
